core/src/test/java/nu/yona/server/messaging/service/MessageServiceTest.java

```diff
@@ -0,0 +1,213 @@
+public class MessageServiceTest extends BaseSpringIntegrationTest
+{
+	@Autowired
+	private UserDeviceRepository userDeviceRepository;
+
+	@Autowired
+	private DeviceAnonymizedRepository deviceAnonymizedRepository;
+
+	@MockBean
+	private UserService mockUserService;
+
+	@MockBean
+	private UserAnonymizedService mockUserAnonymizedService;
+
+	@MockBean
+	private GoalRepository mockGoalRepository;
+
+	@MockBean
+	private MessageSourceRepository mockMessageSourceRepository;
+
+	@MockBean
+	private MessageDestinationRepository mockMessageDestinationRepository;
+
+	@MockBean
+	private MessageRepository mockMessageRepository;
+
+	@MockBean
+	private FirebaseService mockFirebaseService;
+
+	@Autowired
+	private MessageService service;
+
+	private static final String PASSWORD = "password";
+
+	private UUID userAnonId;
+	private UserAnonymized userAnonEntity;
+	private DeviceAnonymized deviceAnonEntity;
+
+	@Before
+	public void setUp()
+	{
+		when(mockUserService.generatePassword()).thenReturn("topSecret");
+
+		// Set up UserAnonymized instance.
+		MessageDestination anonMessageDestinationEntity = MessageDestination
+				.createInstance(PublicKeyUtil.generateKeyPair().getPublic());
+		Set<Goal> goals = new HashSet<>();
+		deviceAnonEntity = DeviceAnonymized.createInstance(0, OperatingSystem.ANDROID, "Unknown", 5,
+				Optional.of("Firebase-12345"));
+		deviceAnonymizedRepository.save(deviceAnonEntity);
+		userAnonEntity = UserAnonymized.createInstance(anonMessageDestinationEntity, goals);
+		userAnonEntity.addDeviceAnonymized(deviceAnonEntity);
+		UserAnonymizedDto userAnon = UserAnonymizedDto.createInstance(userAnonEntity);
+		userAnonId = userAnon.getId();
+
+		when(mockUserAnonymizedService.getUserAnonymized(userAnonId))
+				.thenReturn(UserAnonymizedDto.createInstance(userAnonEntity));
+
+		when(mockMessageDestinationRepository.findById(anonMessageDestinationEntity.getId()))
+				.thenReturn(Optional.of(anonMessageDestinationEntity));
+	}
+
+	@Override
+	protected Map<Class<?>, Repository<?, ?>> getRepositories()
+	{
+		Map<Class<?>, Repository<?, ?>> repositoriesMap = new HashMap<>();
+		repositoriesMap.put(MessageSource.class, mockMessageSourceRepository);
+		repositoriesMap.put(MessageDestination.class, mockMessageDestinationRepository);
+		repositoriesMap.put(Message.class, mockMessageRepository);
+		repositoriesMap.put(DeviceAnonymized.class, deviceAnonymizedRepository);
+		repositoriesMap.put(UserDevice.class, userDeviceRepository);
+		repositoriesMap.put(Goal.class, mockGoalRepository);
+		return repositoriesMap;
+	}
+
+	@Test
+	public void sendMessage_anonymousDestination_sendsFirebaseMessage()
+	{
+		Message message = Mockito.mock(Message.class);
+
+		service.sendMessage(message, UserAnonymizedDto.createInstance(userAnonEntity));
+
+		verify(mockFirebaseService, times(1)).sendMessage("Firebase-12345", message);
+	}
+
+	@Test
+	public void sendMessage_namedDestination_doesNotSendFirebaseMessage()
```

Comment:
I've actually never thought of this. For direct messages, we cannot send a notification, as we do not know the anonymized device ID. We should consider sending them an SMS message for direct messages.

Reply:
Indeed. For now I have implemented that once the message is transferred to the anonymous destination, a notification will be sent. However, that transfer will only happen when the user opens the app, so it is not very useful to then send a notification.

Sending an SMS will have SMS costs, but probably direct messages are not sent that often. Do you propose to remove the mechanism of sending notifications on transfer of a direct message to the anonymous destination and add that an SMS is sent?

Reply:
Created https://jira.yona.nu/browse/YD-605